src/Microsoft.Diagnostics.Monitoring/DiagnosticsMonitor.cs

```diff
@@ -19,359 +19,119 @@ namespace Microsoft.Diagnostics.Monitoring
+            _disposeSource = new CancellationTokenSource();
         }
 
-        public async Task ProcessEvents(int processId, CancellationToken cancellationToken)
+        public Task CurrentProcessingTask => _currentTask;
+
+        public Task<Stream> ProcessEvents(int processId, int duration, CancellationToken cancellationToken)
         {
-            var hasEventPipe = false;
+            cancellationToken.ThrowIfCancellationRequested();
 
-            for (int i = 0; i < 10; ++i)
+            lock (_lock)
             {
-                if (DiagnosticsClient.GetPublishedProcesses().Contains(processId))
+                if (_disposed)
                 {
-                    hasEventPipe = true;
-                    break;
+                    throw new ObjectDisposedException(nameof(DiagnosticsMonitor));
                 }
 
-                cancellationToken.ThrowIfCancellationRequested();
-
-                await Task.Delay(500);
-            }
-
-            if (!hasEventPipe)
-            {
-                _logger.LogInformation("Process id {PID}, does not support event pipe", processId);
-                return;
-            }
-
-            _logger.LogInformation("Listening for event pipe events for {ServiceName} on process id {PID}", _dimValues[1], processId);
+                if (_currentTask != null)
+                {
+                    throw new InvalidOperationException("Only one stream processing is allowed");
+                }
 
-            while (!cancellationToken.IsCancellationRequested)
-            {
                 EventPipeSession session = null;
                 var client = new DiagnosticsClient(processId);
 
                 try
                 {
                     session = client.StartEventPipeSession(_sourceConfig.GetProviders());
                 }
-                catch (EndOfStreamException)
+                catch (EndOfStreamException e)
                 {
-                    break;
+                    throw new InvalidOperationException("End of stream", e);
                 }
-                catch (Exception ex)
+                catch (Exception ex) when (!(ex is OperationCanceledException))
                 {
-                    if (!cancellationToken.IsCancellationRequested)
-                    {
-                        _logger.LogDebug(0, ex, "Failed to start the event pipe session");
-                    }
-
-                    // We can't even start the session, wait until the process boots up again to start another metrics thread
-                    break;
+                    throw new InvalidOperationException("Failed to start the event pipe session", ex);
                 }
 
-                void StopSession()
+                CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(_disposeSource.Token, cancellationToken);
+
+                _currentTask = Task.Run( async () =>
```

Comment:
An alternative to this could be something like:

```
linkedSource.Token.Register(() => {
    StopSession(session);
    _currentTaskSource.TrySetResult(null);
    linkedSource.Dispose();
});
linkedSource.CancelAfter(TimeSpan.FromSeconds(duration));
```

and changing _currentTask to be of type TaskCompletionSource<object>

I think this would avoid creating a Task that executes just to wait for a timeout or cancellation.

Reply:
My mental model for Task is that it is a sub-microsecond operation with a few allocations. On the performance scale we are looking at here (operations that last for seconds or minutes) I wouldn't anticipate that overhead to matter. I'd recommend optimizing for ease of understanding and flexibility to future refactoring. I think using a Task with an async delegate does a good job of that.